pkg/actors/internal/placement/placement.go

```diff
@@ -0,0 +1,360 @@
+				case actorTypes = <-ch:
+					log.Infof("Updating actor types: %v", actorTypes)
+					// TODO: @joshvanl: it is a nonsense to be sending the same actor
+					// types over and over again, *every second*. This should be removed
+					// ASAP.
+				case <-tch.C:
+				case <-ctx.Done():
+					return nil
+				}
+
+				if err := p.sendHost(ctx, actorTypes); err != nil {
+					return err
+				}
+			}
+		},
+		func(ctx context.Context) error {
+			defer p.wg.Wait()
+			defer p.lock.EnsureUnlockTable()
+
+			for {
+				in, err := p.client.Recv(ctx)
+				if err != nil {
+					return err
+				}
+
+				p.handleReceive(ctx, in)
+			}
+		},
+	).Run(ctx)
+
+	p.closed.Store(true)
+	p.lock.EnsureUnlockTable()
+	return err
+}
+
+// LookupActor returns the address of the actor.
+// Placement _must_ be locked before calling this method.
+func (p *placement) LookupActor(ctx context.Context, req *api.LookupActorRequest) (*api.LookupActorResponse, error) {
+	table, ok := p.hashTable.Entries[req.ActorType]
+	if !ok {
+		return nil, ErrActorTypeNotFound
+	}
+
+	host, err := table.GetHost(req.ActorID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &api.LookupActorResponse{
+		Address: host.Name,
+		AppID:   host.AppID,
+		Local:   p.isActorLocal(host.Name, p.hostname, p.port),
+	}, nil
+}
+
+func (p *placement) Ready() bool {
+	return p.client.Ready()
+}
+
+func (p *placement) sendHost(ctx context.Context, actorTypes []string) error {
+	err := p.client.Send(ctx, &v1pb.Host{
+		Name:      p.hostname + ":" + p.port,
+		Entities:  actorTypes,
+		Id:        p.appID,
+		ApiLevel:  20,
+		Namespace: p.namespace,
+	})
+	if err != nil {
+		diag.DefaultMonitoring.ActorStatusReportFailed("send", "status")
+		return err
+	}
+
+	diag.DefaultMonitoring.ActorStatusReported("send")
+
+	return nil
+}
+
+func (p *placement) handleReceive(ctx context.Context, in *v1pb.PlacementOrder) {
+	p.operationLock.Lock()
+	defer p.operationLock.Unlock()
+
+	log.Debugf("Placement order received: %s", in.GetOperation())
+	diag.DefaultMonitoring.ActorPlacementTableOperationReceived(in.GetOperation())
+
+	switch in.GetOperation() {
+	case lockOperation:
+		p.handleLockOperation(ctx)
+	case updateOperation:
+		p.handleUpdateOperation(ctx, in.GetTables())
+	case unlockOperation:
+		p.handleUnlockOperation(ctx)
+	}
+}
+
+func (p *placement) Lock(ctx context.Context) error {
+	select {
+	case <-p.readyCh:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	p.lock.LockLookup()
+	return nil
+}
+
+func (p *placement) Unlock() {
+	p.lock.UnlockLookup()
+}
+
+func (p *placement) handleLockOperation(ctx context.Context) {
+	p.lock.LockTable()
+	lockVersion := p.lockVersion.Add(1)
+
+	clear(p.hashTable.Entries)
+
+	// If we don't receive an unlock in 10 seconds, unlock the table.
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-time.After(time.Second * 10):
+			p.operationLock.Lock()
+			defer p.operationLock.Unlock()
+			if p.updateVersion.Load() < lockVersion {
```

Comment:
A lock version! That's exactly what I was thinking about these days. Thanks for adding it! This is much safer than what we used to have.
Shouldn't we reset the tables again here though? If we only get a lock and an update, but never get the unlock, we will end up with a state that hasn't been properly "committed" by the unlock and might be inconsistent.

Reply:
Done!